README.md
# Sea Search Area Calculator
Author(s): [Bjorn Backeberg](mailto:backeb@gmail.com) (backeb) <br>
Creation date: 15-Sep-2017 <br>
Last updated:  08-Jun-2019 <br>

## Purpose
Use real-time wind (and ocean current coming soon) forecasts to calculate search areas coordinates for persons/objects lost at sea.

This notebook is based on the search area calculator I originally developed for the [OCIMS](https://www.ocims.gov.za/) "Operations at Sea" decision support tool in collaboration with the [NSRI](https://www.nsri.org.za/).

## Dependencies
Manage python installation through [conda](https://docs.conda.io/en/latest/miniconda.html)

Currently using Python 3.7.1

`conda install numpy pandas jupyter basemap xarray netcdf4 dask`  
`conda install -c conda-forge gpxpy`  
`conda install -c conda-forge pynio`  
`conda install -c conda-forge pygrib`  
`conda install -c conda-forge ipyleaflet`  

## Directory tree
```
|- CHANGELOG.md
|- README.md
|- Docs:                                    Relevant documentation
|- app:                                     Main code base
    |- make_gfs_wind_datastack.py:          makes wind forecast datastack for sea_search_area_calculator.py
    |- make_mercator_datastack.py:          makes ocean current forecast datastack for sea_search_area_calculator.py
    |- plot_search_area_map.py:             plot output from sea_search_area_calculator.py
    |- sea_search_area_calculator.py:       calculate search area based on wind and ocean current input data
|- notebooks:                               Development notebooks
    |- sea_search_area_calculator.ipynb:    interactive notebook for sea_search_area_calculator.py
```


CHANGELOG.md
```
Date: Author - Comment  
28 Sep 2017: backeb - added code to plot drift vector on basemap  
03 Oct 2017: backeb - added code get search area radius, note dependency on gpxpy library, to install do   
                      `conda install -c conda-forge gpxpy`  
                      or   
                      `pip install gpxpy --user`  
09 Oct 2017: backeb - added code to get coordinates of the square around the search area  
10 Oct 2017: backeb - rewrote into functions  
01 Nov 2017: backeb - added deltaT, i.e. multiply drift vector by deltaT in hours and recalculate search area  
20 Mar 2018: backeb - rewrote code to do "dynamic" drift vector calculation and include divergence when tdelta >1  
11 Jul 2018: backeb - added opendap call to get real-time winds from GFS, uses xarray and pynio  
                      Note: xarray/pynio engine is buggy when working on large datasets but works fine when  
                      calling point locations. see issue: https://github.com/NCAR/pynio/issues/22  
                      to install pynio do the following:  
                      ```conda create -n pynioEnv python=2.7 anaconda  
                      source activate pynioEnv  
                      conda install numpy pandas jupyter basemap xarray netcdf4 dask  
                      conda install -c ncar pynio```   
02 Aug 2018: backeb - added boolean to get_winds() to use local gfs_winds.nc datastack if available, else call  
                      from GFS opendap  
08 Jun 2019: backeb - UPATE: pynio now works in python 3 - dependencies updated  
                    - Fixed bug in get_winds function and made notebook more interactive  
                    - added widget for input parameters  
                    - writing positions out to pandas dataframe  
                    - conda install -c conda-forge ipyleaflet
                    - added interactive map to notebook
09 Jun 2019: backeb - merged interactive_map_in_notebook branch with master
```


app/plot_search_area_map.py
# coding: utf-8
"""Purpose: plot search area on map

Dependencies: python 2.7

Creation Date: 21 Aug 2018

Author: bjornb <backeb@gmail.com> Bjorn Backeberg

Log: <Date: Author - Comment>
    21 Aug 2018: bjornb - added shoot and equi function to plot circles
"""
from __future__ import print_function

def shoot(lon, lat, azimuth, maxdist=None):
    """Shooter Function
    Original javascript on http://williams.best.vwh.net/gccalc.htm
    Translated to python by Thomas Lecocq

    function calculates the coordinates of the circles for the search areas

    """
    import numpy as np
    glat1 = lat * np.pi / 180.
    glon1 = lon * np.pi / 180.
    s = maxdist / 1.852
    faz = azimuth * np.pi / 180.
 
    EPS= 0.00000000005
    if ((np.abs(np.cos(glat1))<EPS) and not (np.abs(np.sin(faz))<EPS)):
        alert("Only N-S courses are meaningful, starting at a pole!")
 
    a=6378.13/1.852
    f=1/298.257223563
    r = 1 - f
    tu = r * np.tan(glat1)
    sf = np.sin(faz)
    cf = np.cos(faz)
    if (cf==0):
        b=0.
    else:
        b=2. * np.arctan2 (tu, cf)
 
    cu = 1. / np.sqrt(1 + tu * tu)
    su = tu * cu
    sa = cu * sf
    c2a = 1 - sa * sa
    x = 1. + np.sqrt(1. + c2a * (1. / (r * r) - 1.))
    x = (x - 2.) / x
    c = 1. - x
    c = (x * x / 4. + 1.) / c
    d = (0.375 * x * x - 1.) * x
    tu = s / (r * a * c)
    y = tu
    c = y + 1
    while (np.abs (y - c) > EPS):
 
        sy = np.sin(y)
        cy = np.cos(y)
        cz = np.cos(b + y)
        e = 2. * cz * cz - 1.
        c = y
        x = e * cy
        y = e + e - 1.
        y = (((sy * sy * 4. - 3.) * y * cz * d / 6. + x) *
              d / 4. - cz) * sy * d + tu
 
    b = cu * cy * cf - su * sy
    c = r * np.sqrt(sa * sa + b * b)
    d = su * cy + cu * sy * cf
    glat2 = (np.arctan2(d, c) + np.pi) % (2*np.pi) - np.pi
    c = cu * cy - su * sy * cf
    x = np.arctan2(sy * sf, c)
    c = ((-3. * c2a + 4.) * f + 4.) * c2a * f / 16.
    d = ((e * cy * c + cz) * sy * c + y) * sa
    glon2 = ((glon1 + x - (1. - c) * d * f + np.pi) % (2*np.pi)) - np.pi    
 
    baz = (np.arctan2(sa, b) + np.pi) % (2 * np.pi)
 
    glon2 *= 180./np.pi
    glat2 *= 180./np.pi
    baz *= 180./np.pi
 
    return (glon2, glat2, baz)

def equi(centerlon, centerlat, radius):
    """function to calculate the coordinates of a circle of with a radius around lon/lat input point  
    """
    glon1 = centerlon
    glat1 = centerlat
    X = []
    Y = []
    for azimuth in range(0, 360):
        glon2, glat2, baz = shoot(glon1, glat1, azimuth, radius)
        X.append(glon2)
        Y.append(glat2)
    X.append(X[0])
    Y.append(Y[0])
    return X, Y



app/make_gfs_wind_datastack.py
# coding: utf-8
"""Purpose: make datastack of GFS gridded 48 hour wind forecasts and 48 hour hindcast and write to netcdf file

Dependencies: python 3.6, datetime, timedelta, pydap.client, numpy, xarray, os

Creation Date: 27 Jul 2018

Author: bjornb <bbackeberg@csir.co.za> Bjorn Backeberg

Log: <Date: Author - Comment>
    27 Jul 2018: bjornb - added code to get 48 hour forecast and 48 hour hindcast from gfs using xarray and pynio
                        - ISSUE: xarray / pynio bug: https://github.com/NCAR/pynio/issues/22
    02 Aug 2018: bjornb - rewrote code to use pydap instead of pynio (less neat but more robust)
    ___TODO____:        - implement roll over meridian, e.g. lon=ds.lon.values-180; nlon=len(lon); data=np.roll(data, nlon/2, axis=1)
"""


from datetime import datetime, timedelta
from pydap.client import open_url
import numpy as np
import xarray as xr
import pandas as pd
import os

startTime = datetime.now()

# here we set the domain we want 
# [lon-lower-left-corner, lon-upper-right-corner, lat-lower-left-corner, lat-upper-right-corner]
domain = [0, 60, -50, -10]


# here we fetch the latest GFS forcest and write ugrd10m and vgrd10m to u and v
ref_date = datetime.now()
url = "http://nomads.ncep.noaa.gov:9090/dods/gfs_0p25_1hr/gfs{date}/gfs_0p25_1hr_{hour}"
#url = "http://nomads.ncep.noaa.gov/dods/gfs_0p25_1hr/gfs{date}/gfs_0p25_1hr_{hour}"

for z in ["18z", "12z", "06z", "00z"]:
    url_z = url.format(date=ref_date.strftime('%Y%m%d'), hour=z)
    try:
        dataset = open_url(url_z)
        print("adding latest wind forcast " + url_z + " to datastack")
        break
    except:
        pass

# here we subset and load region of interest
lon = dataset['lon'];
lat = dataset['lat']; 
lonIndx = np.squeeze(np.array(np.where((lon >= domain[0]) & (lon <= domain[1]))))
latIndx = np.squeeze(np.array(np.where((lat >= domain[2]) & (lat <= domain[3]))))
ugrd10m = dataset['ugrd10m'][:,slice(latIndx[0],latIndx[-1]),slice(lonIndx[0],lonIndx[-1])]
vgrd10m = dataset['vgrd10m'][:,slice(latIndx[0],latIndx[-1]),slice(lonIndx[0],lonIndx[-1])]
time = dataset['time']
timeUnits = time.attributes['units']
del lon, lat, dataset, url_z

# here we open a xarray.Dataset to which we write ugrd10m and vgrd10m
wds = xr.Dataset()
decode_time = xr.coding.times.decode_cf_datetime # need this guy when setting the time coord in xarray.Dataset
# here we write ugrd10m and vgrd10m to the xarray.Dataset
wds['ugrd10m'] = xr.DataArray(data=ugrd10m.data[0],
        name=ugrd10m.name, dims=ugrd10m.dimensions,
        coords={'time': decode_time(time[:], timeUnits), 'lat': ugrd10m.data[2], 'lon': ugrd10m.data[3]},
        attrs=ugrd10m.attributes)
wds['vgrd10m'] = xr.DataArray(data=vgrd10m.data[0],
        name=vgrd10m.name, dims=vgrd10m.dimensions,
        coords={'time': decode_time(time[:], timeUnits), 'lat': vgrd10m.data[2], 'lon': vgrd10m.data[3]},
        attrs=vgrd10m.attributes)
del ugrd10m, vgrd10m, time


# here we fetch the last 48 hours in 6-hour timesteps
# update ref_date
ref_date = pd.to_datetime(wds.time[0].values) - timedelta(hours=6)

# now we go back and fetch the last 48 hours of GFS forecasts and add to u and v
print("Fetching and adding previous 48 hours of forecasts")
for k in range(0,8):
    url_z = url.format(date=ref_date.strftime('%Y%m%d'), hour=ref_date.strftime('%Hz'))
    dataset = open_url(url_z)
    print ("adding "+url_z+" to datastack")
    ugrd10m = dataset['ugrd10m'][slice(0,6),slice(latIndx[0],latIndx[-1]),slice(lonIndx[0],lonIndx[-1])]
    vgrd10m = dataset['vgrd10m'][slice(0,6),slice(latIndx[0],latIndx[-1]),slice(lonIndx[0],lonIndx[-1])]
    time = dataset['time'][slice(0,6)]
    del dataset, url_z
    # here we write ugrd10m and vgrd10m to the tds xarray.Dataset
    # here we open another xarray.Dataset we use to add to wds
    tds = xr.Dataset()
    tds['ugrd10m'] = xr.DataArray(data=ugrd10m.data[0],
        name=ugrd10m.name, dims=ugrd10m.dimensions,
        coords={'time': decode_time(time[:], timeUnits), 'lat': ugrd10m.data[2], 'lon': ugrd10m.data[3]},
        attrs=ugrd10m.attributes)
    tds['vgrd10m'] = xr.DataArray(data=vgrd10m.data[0],
        name=vgrd10m.name, dims=vgrd10m.dimensions,
        coords={'time': decode_time(time[:], timeUnits), 'lat': vgrd10m.data[2], 'lon': vgrd10m.data[3]},
        attrs=vgrd10m.attributes)
    del ugrd10m, vgrd10m, time
    # here we add tds to the front of wds 
    wds = xr.concat([tds, wds], dim = 'time') 
    del tds
    # update ref_date
    ref_date = pd.to_datetime(wds.time[0].values) - timedelta(hours=6)

del k

# here we write the xarray.Dataset to a netcdf file
fname = "gfs_winds.nc"
try:
    os.remove(fname)
except OSError:
    pass
print("writing datastack to netcdf4 :: "+fname)
wds.to_netcdf(fname, 'w', 'NETCDF4')

print(datetime.now() - startTime)


app/make_mercator_datastack.py
# coding: utf-8
"""Purpose: download latest hourly ocean forecasts from CMEMS GLOBAL-ANALYSIS-FORECAST-PHY-001-024-HOURLY-T-U-V-SSH

Dependencies: python 2.7, motu-client (do 'pip install motu-client'), datetime, timedelta

Creation Date: 21 Aug 2018

Author: bjornb <backeb@gmail.com> Bjorn Backeberg

Log: <Date: Author - Comment>
    21 Aug 2018: bjornb - adapted script from Mostafa Bakhoday-Paskyabi <Mostafa.Bakhoday@nersc.no>
"""
from __future__ import print_function
import os
from datetime import datetime, timedelta

startTime = datetime.now()

# here we define all the bits and pieces that get put in the 'runcommand' variable used to call the motu-client
path2motuClient = '/Users/bjornb/miniconda3/envs/motuEnv/lib/python2.7/site-packages'

usrname = 'bbackeberg'
passwd = 'iaTmwJ7D'

# [lon-lower-left-corner, lon-upper-right-corner, lat-lower-left-corner, lat-upper-right-corner]
domain = [0, 60, -50, -10]

startDate = datetime.now() - timedelta(hours = 48)
startDate = startDate.replace(minute = 30, second = 0, microsecond  = 0)
endDate = datetime.now() + timedelta(hours = 120)
endDate = endDate.replace(minute = 30, second = 0, microsecond  = 0)

# thetao = Temperature in degrees C, zos = SSH in m, uo = Eastward velocity in m/s, vo = Northward velocity in m/s
varList = ['thetao', 'zos', 'uo', 'vo']

# NOTE only surface fields available hourly
depths = [0.493, 0.4942]

path2saveData = os.getcwd()+'/'
fname = 'mercator_ocean.nc'

# create the runcommand string
runcommand = 'python '+path2motuClient+'/motu-client.py --quiet'+ \
        ' --user '+usrname+' --pwd '+passwd+ \
        ' --motu http://nrt.cmems-du.eu/motu-web/Motu'+ \
        ' --service-id GLOBAL_ANALYSIS_FORECAST_PHY_001_024-TDS'+ \
        ' --product-id global-analysis-forecast-phy-001-024-hourly-t-u-v-ssh'+ \
        ' --longitude-min '+str(domain[0])+' --longitude-max '+str(domain[1])+ \
        ' --latitude-min '+str(domain[2])+' --latitude-max '+str(domain[3])+ \
        ' --date-min "'+str(startDate.strftime('%Y-%m-%d %H:%M:%S'))+'" --date-max "'+str(endDate.strftime('%Y-%m-%d %H:%M:%S'))+'"'+ \
        ' --depth-min '+str(depths[0])+' --depth-max '+str(depths[1])+ \
        ' --variable '+varList[0]+' --variable '+varList[1]+' --variable '+varList[2]+' --variable '+varList[3]+ \
        ' --out-dir '+path2saveData+' --out-name '+fname

# run the runcommand, i.e. download the data specified above
print('fetching latest mercator ocean forecast from CMEMS and making datastack')
os.system(runcommand)

print(datetime.now() - startTime)



app/sea_search_area_calculator.py
# coding: utf-8
"""Purpose: calculate search area based on input winds

Dependencies: python 2.7, gpxpy.geo, math, numpy, datetime, pynio

Creation Date: 15 Sep 2017

Author: bjornb <bbackeberg@csir.co.za> Bjorn Backeberg


Log: <Date: Author - Comment>
    28 Sep 2017: bjornb - added code to plot drift vector on basemap
    03 Oct 2017: bjornb - added code get search area radius, note dependency on gpxpy library 
                 to install do "conda install -c conda-forge gpxpy" or "pip install gpxpy --user"
    09 Oct 2017: bjornb - added code to get coordinates of the square around the search area
    10 Oct 2017: bjornb - rewrote into functions
    01 Nov 2017: bjornb - added deltaT, i.e. multiply drift vector by deltaT in hours and recalculate search area
    20 Mar 2018: bjornb - rewrote code to do "dynamic" drift vector calculation and include divergence when tdelta > 1
    11 Jul 2018: bjornb - added opendap call to get real-time winds from GFS, uses xarray and pynio
                          Note: xarray/pynio engine is buggy when working on large datasets but works fine when calling point locations
                                see issue: https://github.com/NCAR/pynio/issues/22
                          to install pynio do the following:
                            `conda create -n pynioEnv python=2.7 anaconda`
                            `source activate pynioEnv`
                            `conda install numpy pandas jupyter basemap xarray netcdf4 dask`
                            `conda install -c ncar pynio`
    02 Aug 2018: bjornb - added boolean to get_winds() to use local gfs_winds.nc datastack if available, else call from GFS opendap
"""
from __future__ import print_function

def get_resultant_coords(lon_start, lat_start, dist, brng):
    """function to calculate new lon/lat coordinate given a starting point, distance and bearing
     
    USAGE
        lon_end, lat_end = get_resultant_coords(lon_start, lat_start, drift, brng)

    INPUT
        lon_start   =	starting longitude in degrees
        lat_start   =	starting latitude in degrees
        dist        =	dist to new position in km
        brng        =	bearing, direction going towards (in 0-360 degrees)
    
    OUTPUT
        lon_end	    =	end longitude in degrees
        lat_end	    =	end latitude in degrees
    """
    import gpxpy.geo
    import math
    R = gpxpy.geo.EARTH_RADIUS/1000 # Radius of the Earth in km
    brng_rad = math.radians(brng) # convert bearing in radians
    
    # convert starting lat/lon to radians
    lat1 = math.radians(lat_start)
    lon1 = math.radians(lon_start)
    
    # get new positions
    lat2 = math.asin( math.sin(lat1)*math.cos(dist/R) + math.cos(lat1)*math.sin(dist/R)*math.cos(brng_rad))
    lon2 = lon1 + math.atan2(math.sin(brng_rad)*math.sin(dist/R)*math.cos(lat1),math.cos(dist/R)-math.sin(lat1)*math.sin(lat2))
    
    # convert back to degrees
    lat_end = math.degrees(lat2)
    lon_end = math.degrees(lon2)
    return lon_end, lat_end

def get_dist(lon1, lat1, lon2, lat2):
    """function that calculates distance between two points
    
    USAGE
        distance = get_dist(lon1, lat1, lon2, lat2)
    
    INPUT
        lon1  =   start longitude
        lat2  =   start latitude
        lon2  =   end longitude
        lat2  =   end latitude
    """
    from math import sin, cos, sqrt, atan2, radians
    import gpxpy.geo

    R = gpxpy.geo.EARTH_RADIUS/1000 # Radius of the Earth in km
    
    # convert to input lon/lat to radians 
    lat1 = radians(lat1)
    lon1 = radians(lon1)
    lat2 = radians(lat2)
    lon2 = radians(lon2)

    dlon = lon2 - lon1
    dlat = lat2 - lat1

    a = sin(dlat / 2)**2 + cos(lat1) * cos(lat2) * sin(dlon / 2)**2
    c = 2 * atan2(sqrt(a), sqrt(1 - a))

    distance = R * c

    return distance

def get_compass_bearing(lon1, lat1, lon2, lat2):
    """function that calculates compass bearing from point 1 to point 2
    
    USAGE
        compass_bearing = get_compass_bearing(lon1, lat1, lon2, lat2)
    
    INPUT
        lon1  =   start longitude
        lat2  =   start latitude
        lon2  =   end longitude
        lat2  =   end latitude
    """
    import math
    import numpy as np

    lat1 = math.radians(lat1)
    lat2 = math.radians(lat2)
    diffLong = math.radians(lon2 - lon1)
    x = math.sin(diffLong) * math.cos(lat2)
    y = math.cos(lat1) * math.sin(lat2) - (math.sin(lat1) * math.cos(lat2) * math.cos(diffLong))
    initial_bearing = math.atan2(x, y)
    initial_bearing = math.degrees(initial_bearing)
    compass_bearing = (initial_bearing + 360) % 360

    return compass_bearing

def get_search_area_coords(lon_pts, lat_pts, sar):
    """function to calculate search area coordinates given a lon/lat points of the drift vectors
            the 4 corners of the search area in lon[0, 1, 2, 3] and lat[0, 1, 2, 3] are always given as:
            lon[0],lat[0] - 'A' the northern most location
            lon[1],lat[1] - 'B' the eastern most location
            lon[2],lat[2] - 'C' the southern most location
            lon[3],lat[3] - 'D' the western most location
    
    USAGE
        lon_sa, lat_sa = get_search_area_coords(lon_pts,lat_pts)
    
    INPUT
        lon_pts	=   longitude points of the drift vector
        lat_pts	=   latitude points of the drift vector
        sar     =   search area radius
    
    OUTPUT
        lon_sa	=   longitudes of search area
        lat_sa	=   latitudes of search area
    """
    from math import sin, cos, radians
    import numpy as np

    # get compass bearing from DATUM to Last Known Position
    if len(lon_pts.shape) > 1:
        compass_bearing = get_compass_bearing(lon_pts[0,-1], lat_pts[0,-1], lon_pts[0,0], lat_pts[0,0])
        # get centre point of furthest long axis
        lonA, latA = get_resultant_coords(lon_pts[0,-1], lat_pts[0,-1], SAR, compass_bearing-180)
        # get centre point of right hand side axis
        lonB, latB = get_resultant_coords(lon_pts[1,-1], lat_pts[1,-1], SAR, compass_bearing-90)
    else:
        compass_bearing = get_compass_bearing(lon_pts[-1], lat_pts[-1], lon_pts[0], lat_pts[0])
        # get centre point of furthest long axis
        lonA, latA = get_resultant_coords(lon_pts[-1], lat_pts[-1], SAR, compass_bearing-180)
        # get centre point of right hand side axis
        lonB, latB = get_resultant_coords(lon_pts[-1], lat_pts[-1], SAR, compass_bearing-90)

    # get distance AB
    distAB = get_dist(lonA, latA, lonB, latB)

    # get angle alpha to calculate length of adjacent and opposite
    compassAB = get_compass_bearing(lonA, latA, lonB, latB)
    alpha = 90-(compass_bearing-compassAB)

    # get 1/2 length of long axis
    long_axis = cos(radians(alpha))*distAB
    # get 1/2 length of short axis
    short_axis = 0.5*(sin(radians(alpha))*distAB+sar)

    # get coordinates of box around search area
    lon = []
    lat = []
    compass_bearing = compass_bearing - 180
    lon1, lat1 = lonA, latA
    for x in range(0, 4):
        compass_bearing = compass_bearing + 90
        if compass_bearing >= 360:
            compass_bearing = compass_bearing - 360
        if x == 0:
            lon2, lat2 = get_resultant_coords(lon1, lat1, long_axis, compass_bearing)
            lon.append(lon2)
            lat.append(lat2)
            lat1 = lat2
            lon1 = lon2
        elif x == 1 or x==3:
            lon2, lat2 = get_resultant_coords(lon1, lat1, 2*short_axis, compass_bearing)
            lon.append(lon2)
            lat.append(lat2)
            lat1 = lat2
            lon1 = lon2
        else:
            lon2, lat2 = get_resultant_coords(lon1, lat1, 2*long_axis, compass_bearing)
            lon.append(lon2)
            lat.append(lat2)
            lat1 = lat2
            lon1 = lon2
     
    # sort search area coordinates so that A is northern most location, and B is western most location
    lat = np.array(lat)
    lon = np.array(lon)
    north = np.where(lat == lat.max())
    east = np.where(lon == lon.max())
    south = np.where(lat == lat.min())
    west = np.where(lon == lon.min())
    lon_sa = [lon[north], lon[east], lon[south], lon[west]]
    lat_sa = [lat[north], lat[east], lat[south], lat[west]]

    lon_sa = np.array(lon_sa)
    lat_sa = np.array(lat_sa)
    
    return lon_sa, lat_sa

def get_time_delta(tLKP, tETA):
    """function to calculate time in hours between input start time and input end time
    
    USAGE
        tdelta = get_time_delta(tLKP, tETA)
    
    INPUT
        tLKP    =   time of last known position, in 'HH:MM' (as string)
        tETA    =   estimated time of arrival of rescue unit, in 'HH:MM' (as string)
    
    OUTPUT
        tdelta  =   time delta in hours
    """
    from datetime import datetime
    
    # define format of input
    FMT = '%H:%M'

    # calcualte time in hours
    tdelta = datetime.strptime(tETA, FMT) - datetime.strptime(tLKP, FMT)
    tdelta = round(float(tdelta.seconds)/60/60) # convert to hours
    
    return tdelta

def get_drift_and_divergence(casualty_type):
    if casualty_type == 'Person in water, state unknown':
        M, N, D = 0.01, 0.08, 40
    elif casualty_type == 'Person in water, with lifejacket':
        M, N, D = 0.02, 0.0, 45
    elif casualty_type == 'Person in water, verticle':
        M, N, D = 0.01, 0.08, 25
    elif casualty_type == 'Person in water, sitting / huddled':
        M, N, D = 0.02, 0.01, 25
    elif casualty_type == 'Person in water, floating on back':
        M, N, D = 0.02, 0.08, 40
    elif casualty_type == 'Liferafts, no ballast pockets, general type':
        M, N, D = 0.05, 0.03, 38
    elif casualty_type == 'Liferafts, no ballast pockets, no canopy, no drogue':
        M, N, D = 0.06, 0.20, 32
    elif casualty_type == 'Liferafts, no ballast pockets, with canopy, with drogue':
        M, N, D = 0.03, 0.0, 38
    elif casualty_type == 'Liferafts, shallow ballast pockets, with canopy, capsized':
        M, N, D = 0.02, -0.10, 12
    elif casualty_type == 'Liferafts, 4 to 6 man, with canopy, with drogue':
        M, N, D = 0.03, 0.04, 20
    elif casualty_type == 'Liferafts, 15 to 25 man, with canopy, with drogue':
        M, N, D = 0.04, 0.08, 15
    elif casualty_type == 'Aviation raft, 4 to 6 man, with canopy, no drogue':
        M, N, D = 0.04, 0.08, 15
    elif casualty_type == 'Sea kayak, with person':
        M, N, D = 0.01, 0.26, 20
    elif casualty_type == 'Homemake wood raft':
        M, N, D = 0.02, 0.18, 25
    elif casualty_type == 'Homemake wood raft, with sail':
        M, N, D = 0.08, 0.18, 45
    elif casualty_type == 'Surfboard with person':
        M, N, D = 0.02, 0.0, 20
    elif casualty_type == 'Windsurfer with person, sail and mast in the water':
        M, N, D = 0.03, 0.1, 16
    elif casualty_type == 'Mono hull, keel, medium displacement':
        M, N, D = 0.04, 0.0, 65
    elif casualty_type == 'Enclosed lifeboat':
        M, N, D = 0.04, -0.08, 30
    elif casualty_type == 'Vessel with outboard motors no drogue':
        M, N, D = 0.07, 0.04, 35
    elif casualty_type == 'Flat bottomed boat, Boston whaler':
        M, N, D = 0.04, 0.04, 30
    elif casualty_type == 'V hull boat':
        M, N, D = 0.03, 0.08, 25
    elif casualty_type == 'Sport fisher, centre open console':
        M, N, D = 0.06, 0.09, 30
    elif casualty_type == 'Commercial fishing vessel type unknown':
        M, N, D = 0.04, 0.02, 65
    elif casualty_type == 'Commercial fishing vessel longline, stern or net':
        M, N, D = 0.04, 0.0, 65
    elif casualty_type == 'Coastal freighter':
        M, N, D = 0.03, 0.0, 65
    elif casualty_type == 'Fishing vessel general debris':
        M, N, D = 0.02, 0.0, 15
    elif casualty_type == 'Cubic meter bait box, loading unknown':
        M, N, D = 0.02, 0.02, 30
    else:
        print('no such casualty type exists')
    return M, N, D


def calc_spd_drctn(u, v):
    """function to calculate speed and direction from u and v components. Return speed in knots
    
    USAGE
        spd, drctn = calc_spd_drctn(u, v)
    
    INPUT
        u = u-component velocity
        v = v-component velocity
        
    OUTPUT
        spd   = speed
        drctn = direction in degrees from north
    """
    import numpy as np
    
    spd = np.sqrt(u**2 + v**2) * 1.94384 # convert to knots
    drctn_trig_to = np.arctan2(u/spd, v/spd)
    drctn_trig_to_degrees = drctn_trig_to * 180/np.pi
    drctn_trig_from_degrees = drctn_trig_to_degrees + 180
    drctn = drctn_trig_from_degrees
    del drctn_trig_to, drctn_trig_to_degrees, drctn_trig_from_degrees
    
    return spd, drctn

def get_winds(lonIn, latIn, timeIn):
    """"function to get real time winds 10m above surface from NCEP for point location
            List of available real time data: http://nomads.ncep.noaa.gov/

        USAGE
            u, v = get_winds(lonIn, latIn, timeIn)

        INPUT
            lonIn  = longitude point location for which to extract the winds
            latIn  = latitude point location for which to extract the winds
            timeIn = time for which to extract the winds, will pull nearest forecast time
                     format = HH:MM

        OUTPUT
            u = u-component winds at 10m above surface for specified lonIn, latIn, timeIn
            v = v-component winds at 10m above surface for specified lonIn, latIn, timeIn
    """
    import numpy as np
    import xarray as xr
    from datetime import datetime
    import os.path

    dateIn = datetime.now()
    dateIn = dateIn.replace(hour = int(timeIn[0:2]), minute = int(timeIn[3:5]), second = 0, microsecond  = 0)

    # try get winds from local OCIMS datastack else pull from NCEP 
    ds = None
    if os.path.isfile("gfs_winds.nc"):
        ds = xr.open_dataset("gfs_winds.nc")\
                .sel(lon = lonIn, lat = latIn, method = "nearest")\
                .sel(time = dateIn.strftime('%Y-%m-%dT%H:%M:%S'), method = "nearest")
        print("Fetched winds for "+str(ds.time.values)+" at "+str(ds.lon.values)+", "+str(ds.lat.values))  
    else:
        print("Locally downloaded datastack not available, pulling data from NCEP")
        for z in ["18z", "12z", "06z", "00z"]:
            url = "http://nomads.ncep.noaa.gov:9090/dods/gfs_0p25_1hr/gfs"+dateIn.strftime('%Y%m%d')+"/gfs_0p25_1hr_"+z
            try:
                ds = xr.open_dataset(url,  engine = 'pynio')\
                                    .sel(lon = lonIn, lat = latIn, method = "nearest")\
                                    .sel(time = dateIn.strftime('%Y-%m-%dT%H:%M:%S'), method = "nearest")
                print(url)
                print("Fetched winds for "+str(ds.time.values)+" at "+str(ds.lon.values)+", "+str(ds.lat.values))  
                break
            except:
                pass

    u = ds.ugrd10m.values
    v = ds.vgrd10m.values
    #time = ds.time

    return u, v


if __name__ == '__main__':
    from numpy import array, zeros
    from datetime import datetime, timedelta
    
    # Last Known Position and Time
    lat_LKP = array(-34.358225)
    lon_LKP = array(18.500566)
    time_LKP = '16:30'
    
    # Rescue Unit ETA
    time_ETA = '19:30'

    CasualtyType = 'Sea kayak, with person'

    # get drift and divergence for casualty type
    M, N, D = get_drift_and_divergence(CasualtyType)
    
    # get time delta
    tdelta = get_time_delta(time_LKP, time_ETA)
    
    # initiatlise some variables
    tdrft = 0.0 # record total drift distance in nautical miles
    if tdelta == 1:
        lon_PTS = zeros([int(tdelta)+1]); lon_PTS[0] = lon_LKP
        lat_PTS = zeros([int(tdelta)+1]); lat_PTS[0] = lat_LKP
    else:
        lon_PTS = zeros([3, int(tdelta)+1]); lon_PTS[:,0] = lon_LKP
        lat_PTS = zeros([3, int(tdelta)+1]); lat_PTS[:,0] = lat_LKP
    
    # CALCULATE DRIFT DISTANCE DYNAMICALLY IN LOOP FOR LENGTH OF TDELTA
    for i in range(0,int(tdelta)):
    
        # sort out time for get_winds()
        timeIn = (datetime.strptime(time_LKP, '%H:%M')+timedelta(hours=i)).strftime('%H:%M')
        # get wind speed and direction - currently pulling from GFS
        # u- and v-component winds in m/s
        if tdelta == 1:
            u10, v10 = get_winds(lon_PTS[i], lat_PTS[i], timeIn)
        else:
            u10, v10 = get_winds(lon_PTS[0, i], lat_PTS[0, i], timeIn)
        # wind speed in knots and direction in degrees coming from
        wnd_spd, wnd_drctn = calc_spd_drctn(u10, v10)
        # wind direction is always coming from, i.e. in this case southerly (coming from the south)
        drctn = wnd_drctn - 180 # convert to going towards

        # drift per hour in nautical miles
        drft = wnd_spd * M + N
        # convert distance from nautical miles to km
        drft = drft * 1.852 # distance in km
    
        # record total drift distance
        tdrft = tdrft + drft
        
        """TODO get ocean currents"""
        """TODO add ocean current as 1:1 drift component to wind drift"""

        if tdelta == 1:
            lon_PTS[i+1], lat_PTS[i+1] = get_resultant_coords(lon_PTS[i], lat_PTS[i], drft, drctn)
        else:
            lon_PTS[0, i+1], lat_PTS[0, i+1] = get_resultant_coords(lon_PTS[0, i], lat_PTS[0, i], drft, drctn)
            lon_PTS[1, i+1], lat_PTS[1, i+1] = get_resultant_coords(lon_PTS[1, i], lat_PTS[1, i], drft, drctn + D)
            lon_PTS[2, i+1], lat_PTS[2, i+1] = get_resultant_coords(lon_PTS[2, i], lat_PTS[2, i], drft, drctn - D)
        
    # define the DATUM
    if tdelta > 1:
        lon_DATUM, lat_DATUM = lon_PTS[0,-1], lat_PTS[0,-1]
    else:
        lon_DATUM, lat_DATUM = lon_PTS[-1], lat_PTS[-1]
        
    """TODO add get_wave call here and include in plot_SAR_map"""

    # define the search area radius as 1/3 of the distance of the total drift
    SAR = tdrft/3
        
    # get search area coordinates
    lon_SA, lat_SA = get_search_area_coords(lon_PTS, lat_PTS, SAR)
    
    print("Suggested search area")
    print("DATUM:   Longitude = "+str(lon_DATUM)+", Latitude = "+str(lat_DATUM))
    print("A:       Longitude = "+str(lon_SA[0])+", Latitude = "+str(lat_SA[0]))
    print("B:       Longitude = "+str(lon_SA[1])+", Latitude = "+str(lat_SA[1]))
    print("C:       Longitude = "+str(lon_SA[2])+", Latitude = "+str(lat_SA[2]))
    print("D:       Longitude = "+str(lon_SA[3])+", Latitude = "+str(lat_SA[3]))


